src/main/java/co/cask/gcp/datastore/source/util/DatastoreSourceQueryUtil.java

```diff
@@ -0,0 +1,368 @@
+
+    List<PathElement> ancestors = config.getAncestor();
+    if (!ancestors.isEmpty()) {
+      filters.add(DatastoreHelper.makeAncestorFilter(constructKey(ancestors, config)).build());
+    }
+
+    if (!filters.isEmpty()) {
+      builder.setFilter(DatastoreHelper.makeAndFilter(filters));
+    }
+
+    return builder.build();
+  }
+
+  /**
+   * Transforms protobuf query instance emitted by the query splitter into {@link Query<Entity>}.
+   *
+   * @param pbQuery protobuf query instance
+   * @param config Hadoop configuration
+   * @return query instance
+   */
+  public static Query<Entity> transformPbQuery(com.google.datastore.v1.Query pbQuery, Configuration config) {
+    EntityQuery.Builder builder = Query.newEntityQueryBuilder()
+      .setNamespace(config.get(DatastoreSourceConstants.CONFIG_NAMESPACE))
+      .setKind(config.get(DatastoreSourceConstants.CONFIG_KIND));
+
+    if (pbQuery.hasFilter()) {
+      builder.setFilter(transformFilterFromPb(pbQuery.getFilter()));
+    }
+
+    return builder.build();
+  }
+
+  /**
+   * Constructs Datastore protobuf key instance based on given list of path elements
+   * and Datastore configuration.
+   *
+   * @param pathElements list of path elements
+   * @param config Datastore configuration
+   * @return Datastore protobuf key instance
+   */
+  private static com.google.datastore.v1.Key constructKey(List<PathElement> pathElements,
+                                                          DatastoreSourceConfig config) {
+    Object[] elements = pathElements.stream()
+      .flatMap(pathElement -> Stream.of(pathElement.getKind(), pathElement.getNameOrId()))
+      .toArray();
+
+    return DatastoreHelper.makeKey(elements)
+      .setPartitionId(getPartitionId(config.getProject(), config.getNamespace()))
+      .build();
+  }
+
+  /**
+   * Appends to the string builder name or id based on given instance type.
+   * Name will be enclosed into the single quotes.
+   *
+   * @param builder string builder
+   * @param kind kind name
+   * @param nameOrId object representing name or id
+   * @return updated string builder
+   */
+  private static StringBuilder appendKindWithNameOrId(StringBuilder builder, String kind, Object nameOrId) {
+    builder.append(kind).append(", ");
+    if (nameOrId instanceof Long) {
+      builder.append(nameOrId);
+    } else {
+      builder.append("'").append(nameOrId).append("'");
+    }
+    return builder;
+  }
+
+  /**
+   * Transforms given value into value holder corresponding to the given field schema type.
+   * If given value is {@link DatastoreSourceQueryUtil#NULL_VALUE}, creates null value holder.
+   *
+   * @param name field name
+   * @param value filter value in string representation
+   * @param schema field schema
+   * @return protobuf value for filter
+   */
+  private static com.google.datastore.v1.Value constructFilterValue(String name, String value, Schema schema) {
+    Schema.Field field = Objects.requireNonNull(schema.getField(name));
```

Comment:
A nicer exception should be thrown, one that includes the field name that is missing.